Python API for Vitals

bitfort opened the pull request:
Summary:
This adds a Python API for vital signs. This keeps the existing semantics of the TorchVital object but extends it with a global map which the python API can use to resolve string names to TorchVital objects.

This map is kept safe through the use of weak_ptrs.

Test Plan:
buck test mode/dev caffe2/test:torch -- --regex vital
buck test //caffe2/aten:vitals

Differential Revision: D26736443

ngimel commented on aten/src/ATen/core/Vitals.h:
```diff
@@ -23,7 +27,7 @@ struct TorchVitalAttr {
 };
 
-struct TorchVital {
+struct TORCH_API TorchVital {
```
are you sure you have to export structs with TORCH_API too?

bitfort replied:
I tried removing the TORCH_API flag, I got:
https://app.circleci.com/pipelines/github/pytorch/pytorch/298367/workflows/0c418b68-728f-4cd0-a79b-89d3ea60ecc8/jobs/12237456
Apr 08 20:06:09 ImportError: /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN2at6vitals9APIVitals8setVitalERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES9_S9_

I think they're needed since this has functions in it. Unless anyone else is more familiar with the build process has a theory?

ngimel commented on aten/src/ATen/core/Vitals.h:
```diff
@@ -40,11 +44,36 @@ struct TorchVital {
+
+private:
+  // weak_ptr keeps semantics where vitals are print-on-delete.
+    std::map<std::string, TorchVital> name_map_;
```
why is this an std::map and not std::unordered_map or, better yet, flat_hash_map? Also, the comment above is unclear, can you please expand it?

bitfort replied:
Moved to flat_hash_map! Removed comment, no longer applies to current code :)

ngimel commented on aten/src/ATen/core/Vitals.cpp:
```diff
@@ -30,5 +33,26 @@ bool torchVitalEnabled() {
+
+  auto iter = name_map_.find(vital_name);
+  TorchVital *vital = nullptr;
+  if (iter == name_map_.end()) {
```
nit: instead of trying to find `vital_name`, you can just unconditionally insert it, and check if insertion was successful? Or is TorchVital(vital_name) constructor expensive?

bitfort replied:
I like the how much cleaner the code would look. I think the issue is that TorchVital destructor prints to stdout; so we really don't want to call the destructor.

ngimel commented on test/test_torch.py:
```diff
@@ -2890,6 +2891,35 @@ def add_neg_dim_tests():
+class TestVitalSigns(TestCase):
+    def test_basic_vitals(self):
+        # Protect current environment variable, if set
+        stash = None
```
given that you'll likely expand the tests, it makes sense to write a context manager that would handle resetting current TORCH_VITAL env, see e.g. `tf32_off` context manager

bitfort replied:
Thanks for the tip! This really cleaned up the code.

ngimel commented on torch/csrc/Module.cpp:
```diff
@@ -175,6 +176,35 @@ static PyObject * THPModule_crashIfATenASAN(PyObject *module, PyObject *arg) {
+  if (at::vitals::VitalsAPI.setVital(name_str, attr_str, value_str)) {
+    Py_RETURN_TRUE;
+  }
+  // Failed to match any supported vital sign, or vital sign was destructed already.
```
this looks like stale comment, setVital will return false only if env variable is not set?

bitfort replied:
Thanks for the catch! Yes.

ngimel commented on test/test_torch.py:
```diff
@@ -2890,6 +2891,35 @@ def add_neg_dim_tests():
+        os.environ['TORCH_VITAL'] = 'ON'
+        self.assertTrue(torch.vitals_enabled())
+        # Ensure we don't segfault a basic call, since we don't have a GET API, it is harder to check.
+        self.assertTrue(torch.set_vital('Dataloader', 'basic_unit_test', 'TEST'))
```
is it possible to test output from python? If not, that's fine, there's cpp test, but please leave a more detailed comment.

bitfort replied:
Currently we don't have a way to test output; and if we added one it could only read a subset of vitals (due to our declare vitals anywhere as a variable approach). I added a comment describing the issue and fixed the existing one. Thanks!

ngimel reviewed (commented):
This looks good, and cool that builds and tests are passing!

robieta commented on aten/src/ATen/core/Vitals.cpp:
```diff
@@ -30,5 +33,26 @@ bool torchVitalEnabled() {
 }
 
+bool APIVitals::setVital(
+    const std::string& vital_name,
```
Curious why this is a string and not an enum.

bitfort replied:
Adopting Bram's pattern. The idea is that people can adhoc declare these, I suppose. I wouldn't be opposed to an enum here.

robieta commented on torch/csrc/Module.cpp:
```diff
@@ -175,6 +176,35 @@ static PyObject * THPModule_crashIfATenASAN(PyObject *module, PyObject *arg) {
   return THPUtils_packInt32(at::_crash_if_asan(THPUtils_unpackInt(arg)));
 }
 
+static PyObject * THPModule_vitals_enabled(PyObject *module, PyObject *args) {
```
These would be much cleaner with PyBind11. (`py_module.def`)

bitfort replied:
Do you have an example? I adopted what I saw used other places in this file.

robieta replied:
Something like https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L934

In your case I think it would be as simple as:
```
py_module.def("vitals_enabled", &at::vitals::torchVitalEnabled);
py_module.def("set_vital", &at::vitals::VitalsAPI.setVital);
```

PyBind11 is great and will figure out all the arguments with template matching.

bitfort replied:
Ah, so basically a different way to register instead of using?
static PyMethodDef TorchMethods[] = { ... }

I'll take a look!

bitfort replied:
"Cannot create a non-constant pointer to member function" -- had to us a lambda.

But wow - this is like magic. it literally did everything for me.

robieta commented on torch/csrc/Module.cpp:
```diff
@@ -175,6 +176,35 @@ static PyObject * THPModule_crashIfATenASAN(PyObject *module, PyObject *arg) {
+  Py_ssize_t num_args = args ? (Py_ssize_t) PyTuple_Size(args) : 0;
+  THPUtils_assert(num_args == 3, "expected exactly 3 arguments");
+
+  if (!at::vitals::torchVitalEnabled()) {
```
Is this needed if `setVital` does its own `torchVitalEnabled()` check?

bitfort replied:
Maybe, kind of, but maybe not? Currently there 3 return values:
- None = vitals not enabled
- True = vitals enabled and value set
- False = vitals enabled and value not set
So technically this is a different return case; but is it helpful to differentiate None and False? Maybe not, since I don't know of anyone who would check it or even under what conditions would cause it other than vitals being disabled. 

Perhaps this does a small amount of optimization by avoiding unpack args when vitals are enabled (reducing overhead)? But the compiler is pretty smart... What do you think?

robieta reviewed (approved):
LGTM.

Commit:
Python API for Vitals (#53238)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/53238

There is a tension for the Vitals design: (1) we want a macro based logging API for C++ and (2) we want a clean python API. Furthermore, we want to this to work with "print on destruction" semantics.

The unfortunate resolution is that there are (2) ways to define vitals:
(1) Use the macros for local use only within C++ - this keeps the semantics people enjoy
(2) For vitals to be used through either C++ or Python, we use a global VitalsAPI object.

Both these go to the same place for the user: printing to stdout as the globals are destructed.

The long history on this diff shows many different ways to try to avoid having 2 different paths... we tried weak pointers & shared pointers, verbose switch cases, etc. Ultimately each ran into an ugly trade-off and this cuts the difference better the alternatives.

Test Plan:
buck test mode/dev caffe2/test:torch -- --regex vital
buck test //caffe2/aten:vitals

Reviewed By: orionr

Differential Revision: D26736443

fbshipit-source-id: 49b9494dd742d196959ae30982ddb788db145e21
